pkg/api/mlflow/dao/repositories/param.go

```diff
@@ -32,48 +57,68 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
+		if err := tx.WithContext(ctx).Clauses(clause.OnConflict{
+			Columns:   []clause.Column{{Name: "run_uuid"}, {Name: "key"}},
+			DoNothing: true,
+		}).CreateInBatches(params, batchSize).Error; err != nil {
+			return eris.Wrap(tx.Error, "error creating params in batch")
 		}
-		if err := r.db.CreateInBatches(dedupedParams, batchSize).Error; err != nil {
-			return eris.Wrap(err, "error creating params in batch after removing duplicates")
+		// if there were ignored conflicts, verify to be exact duplicates
+		if tx.RowsAffected != int64(len(params)) {
+			conflictingParams, err := findConflictingParams(tx, params)
+			if err != nil {
+				return eris.Wrap(err, "error checking for conflicting params")
+			}
+			if len(conflictingParams) > 0 {
+				return ParamConflictError{
+					Message: fmt.Sprintf("conflicting params found: %v", conflictingParams),
+				}
+			}
 		}
+		return nil
+	}); err != nil {
+		return err
 	}
 	return nil
 }
 
-// removeExactMatches will return a new slice of params which excludes exact matches
-func (r ParamRepository) removeExactMatches(_ context.Context, params []models.Param) ([]models.Param, error) {
-	var keys []string
-	paramMap := map[string]models.Param{}
-	for _, param := range params {
-		key := fmt.Sprintf("%v-%v-%v", param.RunID, param.Key, param.Value)
-		keys = append(keys, key)
-		paramMap[key] = param
-	}
-
-	var foundKeys []string
-	if err := r.db.Raw(`
-		select run_uuid || '-' || key || '-' || value
-		from params
-		where run_uuid || '-' || key || '-' || value in ?`, keys).
-		Find(&foundKeys).Error; err != nil {
-		return []models.Param{}, eris.Wrap(err, "problem selecting existing params")
+// findConflictingParams checks if there are conflicting values for the input params. If a key does not
+// yet exist in the db, or if the same key and value already exist for the run, it is not a conflict.
+// If the key already exists for the run but with a different value, it is a conflict. Conflicts are returned.
+func findConflictingParams(tx *gorm.DB, params []models.Param) ([]paramConflict, error) {
+	var conflicts []paramConflict
+	sql := fmt.Sprintf(`WITH new(key, value, run_uuid) AS (VALUES %s)
+		     SELECT current.run_uuid, current.key, current.value as old_value, new.value as new_value
+		     FROM params AS current
+		     INNER JOIN new USING (run_uuid, key)
+		     WHERE new.value != current.value`, makeSqlPlaceholders(params))
+	if err := tx.Raw(sql, makeSqlValues(params)...).
+		Find(&conflicts).Error; err != nil {
+		return nil, eris.Wrap(err, "error fetching params from db")
 	}
+	return conflicts, nil
+}
 
-	for _, foundKey := range foundKeys {
-		delete(paramMap, foundKey)
+// makeSqlPlaceholders collects a string of (?,?,?), (?,?,?), etc, to the length of the params.
+func makeSqlPlaceholders(params []models.Param) string {
+	valuesArray := make([]string, len(params))
+	for i := range params {
+		valuesArray[i] = "(?,?,?)"
 	}
+	return strings.Join(valuesArray, ",")
+}
```

Comment:
I was trying to balance the "values" function which is highly type-specific, but this could just be `makePlaceholders(numberForEach int, total int) string` or the like

Reply:
or maybe return placeholders and values as a tuple from one function

Reply:
I like this second idea!
a generic function that returns both the placeholders string and the values slice sounds like the way to go — once done, please change the other places in our code that use the same pattern (in a separate PR of course)

Reply:
i don't think generic works very well here since we need specific fields from the struct, peculiar to a sql statement

Reply:
the ordered list of fields would be part of the parameters to this generic function — anyway as I commented in another place, this is good enough for this PR